### DISCUSSION TITLE
Get the request body of post or get in request

### DISCUSSION THREAD

**Discussion by mingtiandemeng** [fastapi/fastapi#6938] [Questions] [answered]
I can get the request body of the post method in Flask in this way
```
    data = request.data
   # method post
    if data:
        logger.debug('data   %s', data)

    # method get
    args = request.args or {}
    if args:
        logger.debug('args   %s', args)

    form_data = request.form

    if form_data:
        logger.debug('form   %s', form_data)
```

I can also get the request parameters of the get method in the flash API in the following way
```
    scope = request.scope
    query_string = scope.get("query_string")
```
But I can't find the body in the FastAPI post

What should I do？thanks

**Comment by Mause** [comment, marked as answer]
Please consult the documentation: https://fastapi.tiangolo.com/tutorial/body/

**Comment by klaa97** [comment]
The FastAPI documentation covers how to get the body with FastAPI parameters (and dependencies).

It's also possible to get the Request `body` (as `JSON`, `form`, etc) directly in your method operation (although you won't have automatic documentation); to do so, refer to:
- https://www.starlette.io/requests/#body
- https://fastapi.tiangolo.com/advanced/using-request-directly/

**Comment by mingtiandemeng** [comment]
> FastAPI文档介绍了如何使用FastAPI参数（和依赖项）获取主体。
> 
> 也可以在您的方法操作中直接获取Request `body`（如`JSON`，`form`等）（尽管您不会获得自动文档）；为此，请参考：
> 
> * https://www.starlette.io/requests/#body
> * [https://fastapi.tiangolo.com/advanced/using-request-direct/](https://fastapi.tiangolo.com/advanced/using-request-directly/)

Can this be implemented in middleware

**Comment by mingtiandemeng** [comment]
> > FastAPI文档介绍了如何使用FastAPI参数（和依赖项）获取主体。
> > 也可以在您的方法操作中直接获取Request `body`（如`JSON`，`form`等）（尽管您不会获得自动文档）；为此，请参考：
> > 
> > * https://www.starlette.io/requests/#body
> > * [https://fastapi.tiangolo.com/advanced/using-request-direct/](https://fastapi.tiangolo.com/advanced/using-request-directly/)
> 
> Can this be implemented in middleware

```
@app.middleware("http")
async def add_process_time_header(request: Request, call_next):
    start_time = time.time()
    logger_header = format_request_logger_header(request)
    logger.info(f"{logger_header} Request Begin")
    logger.info(f"{logger_header} query_params {request.query_params}")
    response = await call_next(request)


    jn =  await request.json()
    logger.debug(jn)
    logger.debug('*'*100)
    
    process_time = time.time() - start_time
    response.headers["X-Process-Time"] = str(process_time)
    logger.info(f'{logger_header} Request End time_used {process_time}')

    return response
```
What's wrong with this? When I got to the point of getting JSON, the program got stuck. Can't get the data in JSON

**Comment by mingtiandemeng** [comment]
> 请查阅文档：[https](https://fastapi.tiangolo.com/tutorial/body/) : [//fastapi.tiangolo.com/tutorial/body/](https://fastapi.tiangolo.com/tutorial/body/)



> Please consult the documentation: https://fastapi.tiangolo.com/tutorial/body/


You can take a look at my chat with klaa97, which has the troubles I'm facing now. Thank you.

**Comment by klaa97** [comment]
> > > FastAPI文档介绍了如何使用FastAPI参数（和依赖项）获取主体。
> > > 也可以在您的方法操作中直接获取Request `body`（如`JSON`，`form`等）（尽管您不会获得自动文档）；为此，请参考：
> > > 
> > > * https://www.starlette.io/requests/#body
> > > * [https://fastapi.tiangolo.com/advanced/using-request-direct/](https://fastapi.tiangolo.com/advanced/using-request-directly/)
> > 
> > 
> > Can this be implemented in middleware
> 
> ```
> @app.middleware("http")
> async def add_process_time_header(request: Request, call_next):
>     start_time = time.time()
>     logger_header = format_request_logger_header(request)
>     logger.info(f"{logger_header} Request Begin")
>     logger.info(f"{logger_header} query_params {request.query_params}")
>     response = await call_next(request)
> 
> 
>     jn =  await request.json()
>     logger.debug(jn)
>     logger.debug('*'*100)
>     
>     process_time = time.time() - start_time
>     response.headers["X-Process-Time"] = str(process_time)
>     logger.info(f'{logger_header} Request End time_used {process_time}')
> 
>     return response
> ```
> 
> What's wrong with this? When I got to the point of getting JSON, the program got stuck. Can't get the data in JSON

The problem is due to a known "issue" or design choice in Starlette (check the related issue [here](https://github.com/encode/starlette/issues/495) for more details): consuming the request body in a ASGI Middleware is problematic, therefore I suggest you to check out this doc: https://fastapi.tiangolo.com/advanced/custom-request-and-route/#custom-apiroute-class-in-a-router ; by using a custom APIRoute class, it will possible to retrieve the request body (since you'll have the same interface as in a standard path operation)

**Comment by mingtiandemeng** [comment]
> > > > FastAPI文档介绍了如何使用FastAPI参数（和依赖项）获取主体。
> > > > 也可以在您的方法操作中直接获取Request `body`（如`JSON`，`form`等）（尽管您不会获得自动文档）；为此，请参考：
> > > > 
> > > > * https://www.starlette.io/requests/#body
> > > > * [https://fastapi.tiangolo.com/advanced/using-request-direct/](https://fastapi.tiangolo.com/advanced/using-request-directly/)
> > > 
> > > 
> > > Can this be implemented in middleware
> > 
> > 
> > ```
> > @app.middleware("http")
> > async def add_process_time_header(request: Request, call_next):
> >     start_time = time.time()
> >     logger_header = format_request_logger_header(request)
> >     logger.info(f"{logger_header} Request Begin")
> >     logger.info(f"{logger_header} query_params {request.query_params}")
> >     response = await call_next(request)
> > 
> > 
> >     jn =  await request.json()
> >     logger.debug(jn)
> >     logger.debug('*'*100)
> >     
> >     process_time = time.time() - start_time
> >     response.headers["X-Process-Time"] = str(process_time)
> >     logger.info(f'{logger_header} Request End time_used {process_time}')
> > 
> >     return response
> > ```
> > 
> > 
> > What's wrong with this? When I got to the point of getting JSON, the program got stuck. Can't get the data in JSON
> 
> The problem is due to a known "issue" or design choice in Starlette (check the related issue [here](https://github.com/encode/starlette/issues/495) for more details): consuming the request body in a ASGI Middleware is problematic, therefore I suggest you to check out this doc: https://fastapi.tiangolo.com/advanced/custom-request-and-route/#custom-apiroute-class-in-a-router ; by using a custom APIRoute class, it will possible to retrieve the request body (since you'll have the same interface as in a standard path operation)

thanks

**Comment by klaa97** [comment]
Feel free to close the issue if resolved 👍

**Comment by tiangolo** [comment]
Thanks for the help here @klaa97 and @Mause ! 👏  🙇 

Thanks for reporting back and closing the issue @mingtiandemeng  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
